
DivisionByZeroError

(PHP 7)

Introduction

DivisionByZeroError is thrown when an attempt is made to divide a number by
zero.

Class synopsis

DivisionByZeroError extends ArithmeticError {
/* Inherited methods */
final public string Error::getMessage ( void )
final public Throwable Error::getPrevious ( void )
final public mixed Error::getCode ( void )
final public string Error::getFile ( void )
final public int Error::getLine ( void )
final public array Error::getTrace ( void )
final public string Error::getTraceAsString ( void )
public string Error::__toString ( void )
final private void Error::__clone ( void )
}

User Contributed Notes 2 notes

===============================================================================

<?php
class MathOperation extends Error
{
    protected $n = 10;

    // Try to get the Division by Zero error object and display as Exception
    public function doArithmeticOperation(): string
    {
        try {
            $value = $this->n % 0;
        } catch (DivisionByZeroError $e) {
            return $e->getMessage();
        }
    }
}

$mathOperationObj = new MathOperation();
echo $mathOperationObj->doArithmeticOperation();
===============================================================================

Note that on division by zero 1/0 and module by zero 1%0 an E_WARNING is
triggered first (probably for backward compatibility with PHP5), then the
DivisionByZeroError exception is thrown next.

The result is, for example, that if you set the maximum level of error
detection with error_level(-1) and you also map errors to exception, say
ErrorException, then on division by zero only this latter exception is thrown
reporting "Division by zero". The result is that a code like this:

<?php
// Set a safe environment:
error_reporting(-1);

// Maps errors to ErrorException.
function my_error_handler($errno, $message)
{ throw new ErrorException($message); }

try {
    echo 1/0;
}
catch(ErrorException $e){
    echo "got $e";
}
?>

allows to detect such error in the same way under PHP5 and PHP7, although the
DivisionByZeroError exception is masked off by ErrorException.
